### PR TITLE
feat: added hal/generic/UartGeneric and hal/generic/SynchronousUartGeneric

### PR DESCRIPTION
**Pull request by daantimmer**

### CODE REVIEW

File: hal/generic/SynchronousUartGeneric.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                none,
+                enabled
+            };
+
+            uint32_t baudrate = 115200;
+            RtsFlowControl rtsFlowControl = RtsFlowControl::none;
+        };
+    }
+
+    class SynchronousUartGeneric : public hal::SynchronousSerialCommunication
```
Comment (richardapeters) [inline hal/generic/SynchronousUartGeneric.hpp:33]:
```suggestion
    class SynchronousUartGeneric
        : public hal::SynchronousSerialCommunication
```

File: hal/generic/UartGeneric.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+
```
Comment (richardapeters) [inline hal/generic/UartGeneric.cpp:1]:
```suggestion
```

File: hal/unix/SynchronousUartUnix.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+#ifdef EMIL_OS_DARWIN
+        if (tcsetattr(fileDescriptor, TCSANOW, &settings) != 0)
+            throw std::runtime_error("Could not set port configuration");
+
+        if (darwinSetSpeed)
+        {
+            if (ioctl(fileDescriptor, IOSSIOSPEED, &bitrate) == -1)
+                throw std::runtime_error("Could not set custom baudrate");
+        }
```
Comment (richardapeters) [inline hal/unix/SynchronousUartUnix.cpp:105]: This boolean has been set to true at line 78 and is not modified or used anywhere else
```suggestion
        if (ioctl(fileDescriptor, IOSSIOSPEED, &bitrate) == -1)
            throw std::runtime_error("Could not set custom baudrate");
```

File: hal/unix/SynchronousUartUnix.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if (flowControl)
+            settings.c_cflag |= CRTSCTS;
+        else
+            settings.c_cflag &= ~CRTSCTS;
+
+        settings.c_cflag |= CREAD | CLOCAL;
+
+#ifdef EMIL_OS_DARWIN
+        bool darwinSetSpeed = true;
+#else
```
Comment (richardapeters) [inline hal/unix/SynchronousUartUnix.cpp:79]:
```suggestion
#ifndef EMIL_OS_DARWIN
```

File: hal/generic/SynchronousUartGeneric.hpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        SynchronousUartGeneric(const std::string& portname, const Config& config = {});
+        virtual ~SynchronousUartGeneric() = default;
+
+        void SendData(infra::ConstByteRange data) override;
+        bool ReceiveData(infra::ByteRange data) override;
+
+    private:
+#if defined(EMIL_HAL_WINDOWS)
+        SynchronousUartWindowsConfig serialConfig;
```
Comment (richardapeters) [inline hal/generic/SynchronousUartGeneric.hpp:46]: While you're at it, could you please add a "using Config = ..." in SynchronousUartWindows and use that here?

File: hal/generic/SynchronousUartGeneric.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
+
```
Comment (rjaegers) [inline hal/generic/SynchronousUartGeneric.cpp:1]:
```suggestion
```

**Review by rjaegers — COMMENTED** [review]
The details of ::Config will leak if you want to change the defaults (e.g. baudrate). If you want to truly generalize the instantiation you could create a "generic" config class and add a translation for the specific implementations.

**Pull request merged**